File: validate-binary-search-tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
        return validate(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private boolean validate(TreeNode root, int min_value, int max_value) {
        if (root == null) return true;
        if (root.val < min_value || root.val > max_value) return false;
        return validate(root.left, min_value, root.val - 1) &&
            validate(root.right, root.val + 1, max_value);
    }
}

File: validate-binary-search-tree.cc
#include <cstdlib>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  bool isValidBST(TreeNode *root) {
    return verify(root, INT_MIN, INT_MAX);
  }

  bool verify(TreeNode* root, int low, int high) {
    if (!root) return true;
    return root->val >= low && root->val <= high &&
      verify(root->left, low, root->val - 1) &&
      verify(root->right, root->val + 1, high);
  }
};
